### PR TITLE
Add type definition 

### PR DESCRIPTION
**Pull request by heshan0131**
Part 1 of the work: Added typedef to
- reducers
- actions
- processers
- schemas
- and partially utils

Signed-off-by: Shan He <heshan0131@gmail.com>

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -49,7 +49,8 @@
     "docs": "babel-node ./scripts/documentation.js",
     "example-version": "babel-node ./scripts/edit-version.js",
     "prettier-all": "prettier --write '{src,examples,test}/**/*.js'",
-    "lint": "eslint src test webpack examples/*.js examples/**/src website/*.js website/src --fix",
+    "lint": "yarn typescript && eslint src test webpack examples/*.js examples/**/src website/*.js website/src --fix",
+    "typescript": "tsc",
```
Comment (ibgreen) [inline package.json:53]: To be clear we can write `tsc --noEmit` - though we do have the noEmit setting in the tsconfig, I still think it is worth having it here to emphasize that we are just type checking.

File: src/actions/actions.d.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import ActionTypes from 'constants/action-types';
+import {ParsedConfig} from '../schemas';
+import {RGBColor} from 'reducers/types';
+import {Bounds} from 'reducers/map-state-updaters';
+import {MapInfo} from 'reducers/vis-state-updaters';
+
+/**
+ * Input dataest parsed to addDataToMap
+ */
+export type ProtoDataset = {
```
Comment (macrigiuseppe) [inline src/actions/actions.d.ts:30]: why protodataset and not simply Dataset?

File: src/actions/actions.js [hunk 3]
```diff
@@ -149,14 +149,16 @@ export const receiveMapConfig = createAction(ActionTypes.RECEIVE_MAP_CONFIG, (co
 /**
  * Initialize kepler.gl reducer. It is used to pass in `mapboxApiAccessToken` to `mapStyle` reducer.
  * @memberof main
- * @param {Object} payload
- * @param {string} payload.mapboxApiAccessToken - mapboxApiAccessToken to be saved to mapStyle reducer
- * @param {string} payload.mapboxApiUrl - mapboxApiUrl to be saved to mapStyle reducer.
- * @param {Boolean} payload.mapStylesReplaceDefault - mapStylesReplaceDefault to be saved to mapStyle reducer
+ * @param {object} payload
+ * @param payload.mapboxApiAccessToken - mapboxApiAccessToken to be saved to mapStyle reducer
```
Comment (macrigiuseppe) [inline src/actions/actions.js:153]: why did you remove types here?

File: src/actions/identity-actions.d.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import ActionTypes from 'constants/action-types';
+
+export function registerEntry(entry: {
```
Comment (macrigiuseppe) [inline src/actions/identity-actions.d.ts:23]: can you create a a type for the entry inptu?
```
type RegistryEntity {
  id:...
  mint:...
...
}
```

File: src/actions/actions.js (hunk 3 above)
Comment (macrigiuseppe) [reply src/actions/actions.js:153]: never mind i see the ```@type```

File: src/actions/actions.d.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+export type AddDaataToMapOptions = {
+  centerMap?: boolean;
+  readOnly?: boolean;
+  keepExistingConfig?: boolean;
+};
+
+export type AddDataToMaoPayload = {
+  datasets: ProtoDataset[];
+  options?: AddDaataToMapOptions;
```
Comment (macrigiuseppe) [inline src/actions/actions.d.ts:58]: type in AddDaat

File: src/actions/actions.d.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  config?: ParsedConfig;
+  info?: Partial<MapInfo>
+};
+
+export function addDataToMap(
+  data: AddDataToMaoPayload
+): {type: ActionTypes.ADD_DATA_TO_MAP; payload: AddDataToMaoPayload};
```
Comment (macrigiuseppe) [inline src/actions/actions.d.ts:65]: typo AddDataToMaoPayload

File: src/actions/map-style-actions.d.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+};
+
+export function mapConfigChange(
+  mapStyle: MapConfigChangeUpdaterAction['payload']
+): Merge<MapConfigChangeUpdaterAction, {type: ActionTypes.MAP_CONFIG_CHANGE}>;
+
+/** REQUEST_MAP_STYLES */
+export type RequestMapStylesUpdaterAction = {
+  payload: {
+    [key: string]: {
```
Comment (macrigiuseppe) [inline src/actions/map-style-actions.d.ts:61]: is this BaseMapStyle?

File: src/actions/provider-actions.d.ts [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+  payload: ExportFileToCloudPayload;
+};
+
+/** EXPORT_FILE_SUCCESS */
+export type ExportFileSuccessPayload = {
+  response: any;
+  provider: Provider;
+  options?: ExportFileOptions;
+  onSuccess?: any;
+  closeModal?: booleam;
```
Comment (macrigiuseppe) [inline src/actions/provider-actions.d.ts:68]: typo in booleam

File: src/actions/map-style-actions.d.ts [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+  };
+};
+export function requestMapStyles(
+  mapStyles: RequestMapStylesUpdaterAction['payload']
+): Merge<RequestMapStylesUpdaterAction, {type: ActionTypes.REQUEST_MAP_STYLES}>;
+
+/** LOAD_MAP_STYLES */
+export type LoadMapStylesUpdaterAction = {
+  payload: {
+    [id: string]: BaseMapStyle;
```
Comment (macrigiuseppe) [inline src/actions/map-style-actions.d.ts:77]: same here

File: src/actions/provider-actions.d.ts [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+  type: ProviderActionTypes.EXPORT_FILE_SUCCESS;
+  payload: ExportFileSuccessPayload;
+};
+
+/** EXPORT_FILE_ERROR */
+export type ExportFileErrorPayload = {
+  error: any;
+  provider: Provider;
+  options?: Options;
+  onError?: any;
```
Comment (macrigiuseppe) [inline src/actions/provider-actions.d.ts:82]: (error: object) => void

Comment (macrigiuseppe) [reply src/actions/provider-actions.d.ts:82]: maybe it would be better to create a type ErrorCallback since you are using it in few places

File: src/actions/provider-actions.d.ts [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+): {
+  type: ProviderActionTypes.POST_SAVE_LOAD_SUCCESS;
+  payload: PostSaveLoadSuccessPayload;
+};
+
+/** LOAD_CLOUD_MAP */
+export type LoadCloudMapPayload = {
+  loadParams: any;
+  provider: string;
+  onSuccess?: any;
```
Comment (macrigiuseppe) [inline src/actions/provider-actions.d.ts:119]: same for this one

File: src/actions/provider-actions.d.ts [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+  onError?: any;
```
Comment (macrigiuseppe) [inline src/actions/provider-actions.d.ts:120]: ErrorCallback

File: src/actions/provider-actions.d.ts [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+  type: ProviderActionTypes.LOAD_CLOUD_MAP;
+  payload: LoadCloudMapPayload;
+};
+
+/** LOAD_CLOUD_MAP_SUCCESS */
+export type LoadCloudMapSuccessPayload = {
+  response: any;
+  loadParams: any;
+  provider: string;
+  onSuccess?: any;
```
Comment (macrigiuseppe) [inline src/actions/provider-actions.d.ts:134]: here

File: src/actions/provider-actions.d.ts [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+): {
+  type: ProviderActionTypes.LOAD_CLOUD_MAP_SUCCESS;
+  payload: LoadCloudMapSuccessPayload;
+};
+
+/** LOAD_CLOUD_MAP_ERROR */
+export type LoadCloudMapErrorPayload = {
+  error: any;
+  provider: stirng;
+  onError?: any;
```
Comment (macrigiuseppe) [inline src/actions/provider-actions.d.ts:148]: here

File: src/actions/vis-state-actions.d.ts [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+import {ProtoDataset, AddDataToMaoPayload} from '../actions/actions';
+import {ParsedConfig} from '../schemas';
+import {FileCacheItem} from '../processors/file-handler';
+import {Layer, LayerConfig, LayerVisConfig} from 'layers';
+import {Feature, InteractionConfig} from 'reducers/vis-state-updaters';
+import {ValueOf, Merge} from '../reducers/types';
+
+export type LayerConfigChangeUpdaterAction = {
+  oldLayer: Layer;
+  newConfig: Partiel<LayerConfig>;
```
Comment (macrigiuseppe) [inline src/actions/vis-state-actions.d.ts:32]: is this a typo? should it be Partial? I am not sure

File: src/constants/action-types.js [hunk 16]
```diff
@@ -51,130 +50,120 @@ import {ACTION_PREFIX} from './default-settings';
  *
  * export default createStore(reducers, {}, applyMiddleware(taskMiddleware))
  */
-const ActionTypes = keyMirror({
+const ActionTypes = {
```
Comment (macrigiuseppe) [inline src/constants/action-types.js:53]: this could be achieved by creating a helper that takes the map object and add the prefix to each and every property

```
Object.keys(types).reduce((acc, key) => ({
  [key]: `${ACTION_PREFIX}${types[key]}`
}), {})
```

I mean you. already did the work so it is not a big deal but something to consider

File: src/layers/layer-factory.d.ts [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+  range: [number, number];
+  step: number;
+};
+
+export type VisConfigBoolean = VisConfig & {
+  type: 'boolean';
+  defaultValue: boolean;
+};
+
+export type VisConfigSelection = VisConfig<string> & {
```
[the remark this answers is not in the document]
Comment (macrigiuseppe) [reply src/layers/layer-factory.d.ts:57]: I don't think VisConfig is a generic, this may be a typo

File: src/actions/actions.d.ts (hunk 2 above)
Comment (heshan0131) [reply src/actions/actions.d.ts:30]: this is the input data that user passed to addDAtaToMap, it then gets converted to dataset

File: src/reducers/vis-state-merger.d.ts [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+import {ParsedConfig} from 'schemas';
+
+export function mergeAnimationConfig(state: VisState, animation: ParsedConfig['visState']['animationConfig']): VisState;
+export function mergeFilters(state: VisState, filters: ParsedConfig['visState']['filters']): VisState;
+export function mergeInteractions(state: VisState, interactionConfig: ParsedConfig['visState']['interactionConfig']): VisState;
+export function mergeLayerBlending(state: VisState, layerBlending: ParsedConfig['visState']['layerBlending']): VisState;
+export function mergeLayers(state: VisState, layers: ParsedConfig['visState']['layers']): VisState;
+export function mergeSplitMaps(state: VisState, splitMaps: ParsedConfig['visState']['splitMaps']): VisState;
+
+export function mergeInteractionTooltipConfig(state: VisState)
```
Comment (macrigiuseppe) [inline src/reducers/vis-state-merger.d.ts:11]: empty line is missing in this file

File: src/schemas/index.d.ts [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+  KeplerGLSchema,
+  ParsedConfig,
+  ParsedDataset,
+  SavedConfigV1,
+  SavedDatasetV1,
+  SavedMap
+} from './schema-manager';
+ 
+export const KeplerGlSchema: KeplerGLSchema;
+export default KeplerGlSchema
```
Comment (macrigiuseppe) [inline src/schemas/index.d.ts:14]: empty line

File: src/actions/map-style-actions.d.ts (hunk 7 above)
Comment (heshan0131) [reply src/actions/map-style-actions.d.ts:61]: this is the request payload, BaseMapStyle is the merged result of this request.

File: src/layers/layer-factory.d.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (heshan0131) [reply src/layers/layer-factory.d.ts:57]: it's a typo, good catch
